File: Behavioral/Iterator/java/Iterator.java
interface Collection {
    
        public void add(double x);
        public int size();
        public Iterator createIterator();
    }
    
    interface Iterator {
    
        public void first();
        public void next();
        public boolean isDone();
        public double currentElement();
    }
    
    class List implements Collection{
    
        private double [] elements;
        private int free; // Marks the first empty slot in the array
    
        public List(int size){
            //Creates the iterable list.
            elements = new double[size];
            free = 0;
        }
    
        public void add(double x){
    
            elements[free] = x;
            free++;
        }
    
        public int size(){
    
            return free;
        }
        
        public double getElement(int index){
    
            return elements[index];
        }
    
        public Iterator createIterator(){
    
            return new ListIterator(this);
            //return new ReverseIterator4(this);
        }
    }
    
    class ListIterator implements Iterator {
    
        private List lista;
        private int index;
    
        public ListIterator(List lista){
    
            this.lista = lista;
        }
    
        public void first(){
    
            index = 0;
        }
    
        public void next(){
    
            index++;
        }
    
        public boolean isDone(){
            //returns true if the index is the last element.
            return index == lista.size();
        }
    
        public double currentElement(){
            
            return lista.getElement(index);
        }
    }
    
    class ReverseIterator implements Iterator{
    
        private List list;
        private int index;
    
        public ReverseIterator(List list){
    
            this.list = list;
        }
    
        public void first(){
    
            index = list.size() - 1;
        }
    
        public void next(){
    
            index--;
        }
    
        public boolean isDone(){
    
            return index < 0;
        }
    
        public double currentElement(){
    
            return list.getElement(index);
        }
    }
    
    public class Iterator {
    
        public static void printList(Collection colecao){
    
            System.out.print("Elements: ");
    
            Iterator it = colecao.createIterator();
            
            for(it.first(); !it.isDone(); it.next()){
                 
                System.out.print(it.currentElement() + " ");
            }
    
            System.out.println();
        }
    
        public static void main(String [] args){
    
            double [] values = { 1.0, 5.5, 3.2, 2.3, 9.8 };	
            
            Collection colecao = new List(10);
    
            for(double d : values) colecao.add(d);
    
            printList(colecao);
        }
    }
    
    

File: Behavioral/Iterator/c++11/iterator.cpp
#include <iostream>

//For C++11
using namespace std;


class Iterator {
    public:
        virtual void first() = 0;
        virtual void next() = 0;
        virtual bool isDone() = 0;
        virtual double currentElement() = 0;
};

class Collection {
    public:
        virtual void add(double x)  = 0;
        virtual int size() = 0;
        virtual Iterator* createIterator() = 0;
};

class List:public Collection{
    private:
        double * elem;
        int capacity;
        int free;
    public:
        List(int capacity): free{0},capacity{capacity}, elem{new double[capacity]} {}
        void add(double x){
            if (free<capacity)
                elem[free++] = x;
            else
                cout << "Unable to add list is already full!";
        }
        int size(){
            return free;
        }
        double getElement(int index){
            return elem[index];
        }
        Iterator* createIterator();
        
};


class ListIterator: public Iterator{
    List * list_instance;
    private:
        int index;
    public:
        ListIterator(List * list_inst): index{0}, list_instance{list_inst} {};
        void first(){
            index = 0;
        }
        void next(){
            index++;
        }
        bool isDone(){
            return index == list_instance->size();
        }
        double currentElement(){
            return list_instance->getElement(index);
        }
};


Iterator * List::createIterator(){
    return new ListIterator(this);
}

int main(){
    double values[] = {1.0, 5.5, 3.2, 2.3, 9.8};
    List * list = new List(5);
    for (double x: values)
        list->add(x);
    Iterator* it = list->createIterator();
    for (it->first(); !it->isDone(); it->next()){
        cout << it->currentElement() << " ";
    }
    cout << endl;
    return 0;
}

